man/pw.1
.TH pw 1 "2016 Sept 01" Pzl "pw"

.SH NAME
pw - stores, retrieves, generates, and synchronizes passwords securely

.SH SYNOPSIS
.B pw
[ 
.I COMMAND
] [ 
.I OPTIONS
]... [ 
.I ARGS
]...

.SH DESCRIPTION

.B pw 
is a very simple password store that stores passwords inside a
.BR gpg2 (1)
encrypted directory. The
.B pw
utility provides a series of commands for manipulating the password store,
allowing the user to add, remove, edit, synchronize, generate, and manipulate
passwords.

If no COMMAND is specified, COMMAND defaults to 
.BR ls .
Otherwise COMMAND must be one of the valid commands listed below.

There is a corresponding bash completion script for use with tab completing
password names in
.BR bash (1).

.SH COMMANDS

.TP
\fBls\fP \fIsubfolder\fP
List names of passwords inside the tree at
.I subfolder
by using the
.BR tree (1)
program. This command is alternatively named \fBlist\fP.
.TP
\fBgrep\fP \fIsearch-string\fP
Searches inside each decrypted password file for \fIsearch-string\fP, and displays line
containing matched string along with filename. Uses
.BR grep (1)
for matching. Make use of the \fIGREP_OPTIONS\fP environment variable to set particular
options.
.TP
\fBfind\fP \fIpass-names\fP...
List names of passwords inside the tree that match \fIpass-names\fP by using the
.BR tree (1)
program. This command is alternatively named \fBsearch\fP.
.TP
\fBshow\fP [ \fI--clip\fP[=\fIline-number\fP], \fI-c\fP[\fIline-number\fP] ] \fIpass-name\fP
Decrypt and print a password named \fIpass-name\fP. If \fI--clip\fP or \fI-c\fP
is specified, do not print the password but instead copy the first (or otherwise specified)
line to the clipboard using
.BR xclip (1)
and then restore the clipboard after 45 (or \fIPASSWORD_STORE_CLIP_TIME\fP) seconds.
.TP
\fBinsert\fP [ \fI--multiline\fP, \fI-m\fP ] [ \fI--force\fP, \fI-f\fP ] \fIpass-name\fP
Insert a new password into the password store called \fIpass-name\fP. This will
read the new password from standard in. If \fI--echo\fP or \fI-e\fP is \fInot\fP specified,
disable keyboard echo when the password is entered and confirm the password by asking
for it twice. If \fI--multiline\fP or \fI-m\fP is specified, lines will be read until
EOF or Ctrl+D is reached. Otherwise, only a single line from standard in is read. Prompt
before overwriting an existing password, unless \fI--force\fP or \fI-f\fP is specified. This
command is alternatively named \fBadd\fP.
.TP
\fBedit\fP \fIpass-name\fP
Insert a new password or edit an existing password using the default text editor specified
by the environment variable \fIEDITOR\fP or using
.BR vi (1)
as a fallback. This mode makes use of temporary files for editing, but care is taken to
ensure that temporary files are created in \fI/dev/shm\fP in order to avoid writing to
difficult-to-erase disk sectors. If \fI/dev/shm\fP is not accessible, fallback to
the ordinary \fITMPDIR\fP location, and print a warning.
.TP
\fBgenerate\fP [ \fI--no-symbols\fP, \fI-n\fP ] [ \fI--clip\fP, \fI-c\fP ] [ \fI--in-place\fP, \fI-i\fP | \fI--force\fP, \fI-f\fP ] \fIpass-name [pass-length]\fP
Generate a new password using
.BR pwgen (1)
of length \fIpass-length\fP (or \fIPASSWORD_STORE_GENERATED_LENGTH\fP if unspecified)
and insert into \fIpass-name\fP. If \fI--no-symbols\fP or \fI-n\fP
is specified, do not use any non-alphanumeric characters in the generated password.
If \fI--clip\fP or \fI-c\fP is specified, do not print the password but instead copy
it to the clipboard using
.BR xclip (1)
and then restore the clipboard after 45 (or \fIPASSWORD_STORE_CLIP_TIME\fP) seconds.
Prompt before overwriting an existing password,
unless \fI--force\fP or \fI-f\fP is specified. If \fI--in-place\fP or \fI-i\fP is
specified, do not interactively prompt, and only replace the first line of the password
file with the new generated password, keeping the remainder of the file intact.
.TP
\fBrm\fP [ \fI--recursive\fP, \fI-r\fP ] [ \fI--force\fP, \fI-f\fP ] \fIpass-name\fP
Remove the password named \fIpass-name\fP from the password store. This command is
alternatively named \fBremove\fP or \fBdelete\fP. If \fI--recursive\fP or \fI-r\fP
is specified, delete pass-name recursively if it is a directory. If \fI--force\fP
or \fI-f\fP is specified, do not interactively prompt before removal.
.TP
\fBmv\fP [ \fI--force\fP, \fI-f\fP ] \fIold-path\fP \fInew-path\fP
Renames the password or directory named \fIold-path\fP to \fInew-path\fP. This
command is alternatively named \fBrename\fP. If \fI--force\fP is specified,
silently overwrite \fInew-path\fP if it exists. If \fInew-path\fP ends in a
trailing \fI/\fP, it is always treated as a directory. Passwords are selectively
reencrypted to the corresponding keys of their new destination.
.TP
\fBcp\fP [ \fI--force\fP, \fI-f\fP ] \fIold-path\fP \fInew-path\fP
Copies the password or directory named \fIold-path\fP to \fInew-path\fP. This
command is alternatively named \fBcopy\fP. If \fI--force\fP is specified,
silently overwrite \fInew-path\fP if it exists. If \fInew-path\fP ends in a
trailing \fI/\fP, it is always treated as a directory. Passwords are selectively
reencrypted to the corresponding keys of their new destination.
.TPq
\fBhelp\fP
Show usage message.
.TP
\fBversion\fP
Show version information.

.SH SIMPLE EXAMPLES

.TP
List existing passwords in store
.B pw 
.br
\[u251C]\[u2500]\[u2500] Business 
.br
\[u2502]   \[u251C]\[u2500]\[u2500] some-silly-business-site.com 
.br
\[u2502]   \[u2514]\[u2500]\[u2500] another-business-site.net 
.br
\[u251C]\[u2500]\[u2500] Email 
.br
\[u2502]   \[u251C]\[u2500]\[u2500] donenfeld.com 
.br
\[u2502]   \[u2514]\[u2500]\[u2500] zx2c4.com 
.br
\[u2514]\[u2500]\[u2500] France 
.br
    \[u251C]\[u2500]\[u2500] bank 
.br
    \[u251C]\[u2500]\[u2500] freebox 
.br
    \[u2514]\[u2500]\[u2500] mobilephone  
.br

.br
Alternatively, "\fBpw ls\fP".
.TP
Find existing passwords in store that match .com
.B pw find .com
.br
Search Terms: .com
.br
\[u251C]\[u2500]\[u2500] Business 
.br
\[u2502]   \[u251C]\[u2500]\[u2500] some-silly-business-site.com 
.br
\[u2514]\[u2500]\[u2500] Email 
.br
    \[u251C]\[u2500]\[u2500] donenfeld.com 
.br
    \[u2514]\[u2500]\[u2500] zx2c4.com 
.br

.br
Alternatively, "\fBpw search .com\fP".
.TP
Show existing password
.B pw Email/zx2c4.com 
.br
sup3rh4x3rizmynam3 
.TP
Copy existing password to clipboard
.B pw -c Email/zx2c4.com 
.br
Copied Email/jason@zx2c4.com to clipboard. Will clear in 45 seconds.
.TP
Add password to store
.B pw insert Business/cheese-whiz-factory 
.br
Enter password for Business/cheese-whiz-factory: omg so much cheese what am i gonna do
.TP
Add multiline password to store 
.B pw insert -m Business/cheese-whiz-factory 
.br
Enter contents of Business/cheese-whiz-factory and press Ctrl+D when finished: 
.br
 
.br
Hey this is my 
.br
awesome 
.br
multi 
.br
line 
.br
passworrrrrrrrd. 
.br
^D 
.TP
Generate new password
.B pw generate Email/jasondonenfeld.com 15 
.br
The generated password to Email/jasondonenfeld.com is: 
.br
$(-QF&Q=IN2nFBx
.TP
Generate new alphanumeric password
.B pw generate -n Email/jasondonenfeld.com 12
.br
The generated password to Email/jasondonenfeld.com is: 
.br
YqFsMkBeO6di
.TP
Generate new password and copy it to the clipboard
.B pw generate -c Email/jasondonenfeld.com 19
.br
Copied Email/jasondonenfeld.com to clipboard. Will clear in 45 seconds.
.TP
Remove password from store
.B pw remove Business/cheese-whiz-factory 
.br
rm: remove regular file \[u2018]/home/zx2c4/.password-store/Business/cheese-whiz-factory.gpg\[u2019]? y 
.br
removed \[u2018]/home/zx2c4/.password-store/Business/cheese-whiz-factory.gpg\[u2019]

.SH FILES

.TP
.B $XDG_DATA_HOME/pw
The default password storage directory.

.SH ENVIRONMENT VARIABLES

.TP
.I XDG_DATA_HOME
Overrides the default password storage directory.
.I PASSWORD_STORE_GPG_OPTS
Additional options to be passed to all invocations of GPG.
.TP
.I PASSWORD_STORE_X_SELECTION
Overrides the selection passed to \fBxclip\fP, by default \fIclipboard\fP. See
.BR xclip (1)
for more info.
.TP
.I PASSWORD_STORE_CLIP_TIME
Specifies the number of seconds to wait before restoring the clipboard, by default
\fI45\fP seconds.
.TP
.I PASSWORD_STORE_UMASK
Sets the umask of all files modified by pw, by default \fI077\fP.
.TP
.I PASSWORD_STORE_GENERATED_LENGTH
The default password length if the \fIpass-length\fP parameter to \fBgenerate\fP
is unspecified.
.TP
.I EDITOR
The location of the text editor used by \fBedit\fP.
.SH SEE ALSO
.BR gpg2 (1),
.BR pwgen (1),
.BR git (1),
.BR xclip (1).

.SH AUTHOR
.B pw
was written by
.MT Jason@zx2c4.com
Jason A. Donenfeld
.ME .
based on
.B pass
originall written by
.MT Jason@zx2c4.com
Jason A. Donenfeld
.ME .
For updates and more information, see
.UR https://\:github.com/pzl/pw
World Wide Web
.UE .

.SH COPYING
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


src/pw.sh
#!/usr/bin/env bash

# Copyright (C) 2012 - 2014 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
# Copyright (C) 2016 Dan Panzarella <alsoelp@gmail.com>.
# This file is licensed under the GPLv2. Please see LICENSE for more information.

umask "${PASSWORD_STORE_UMASK:-077}"
set -o pipefail
#set -x
set -e

GPG_OPTS=( $PW_GPG_OPTS "--quiet" "--yes" "--compress-algo=none" "--no-encrypt-to" )
GPG="gpg"
export GPG_TTY="${GPG_TTY:-$(tty 2>/dev/null)}"
which gpg2 &>/dev/null && GPG="gpg2"
[[ -n $GPG_AGENT_INFO || $GPG == "gpg2" ]] && GPG_OPTS+=( "--batch" )

CLIP_TIME="${CLIP_TIME:-15}"
GENERATED_LENGTH="${PASSWORD_STORE_GENERATED_LENGTH:-25}"


FLNAME="pw"
TMPDIR="/dev/shm"
ENCDIR="${XDG_DATA_HOME:-$HOME/.local/share}"
EDITOR="${EDITOR:-vi}"


#
# BEGIN helper functions
#

approve() {
	local response
	if [ -t 0 ]; then
		read -r -p "$1 [y/N] " response
		[[ $response == [yY] ]] || return 1
		return 0
	else
		response="$(pinentry-gtk-2 <<EOD
setdesc $1
settitle pw
setok Yes
setcancel No
confirm
EOD
)"
		echo "$response" | grep "^ERR.*cancelled" && return 1
		return 0
	fi
}

say() {
	local soft=0
	if [[ "$1" == "softly" ]]; then
		soft=1
		shift
	fi

	if [ -t 0 ]; then
		echo "$@"
	else
		if [[ $soft -eq 1 ]]; then
			notify-send "pw" "$@"
		else
			pinentry-gtk-2 <<EOD
setdesc $@
settitle pw
message
EOD
		fi
	fi
}

die() {
	if [ -t 0 ]; then
		echo "$@" >&2
	else
		pinentry-gtk-2 <<EOD
setdesc $@
settitle pw error
message
EOD
	fi
	cleanup
	exit 1
}

askkey() {
	if [ -t 0 ]; then
		read -s -p "$1: " USER_PW || return 1
		printf "\n"
		#echo -ne "\033[1G\033[2K" #erases pass line
	else
		local answers
		answers="$( pinentry-gnome3 <<EOD
setdesc $1
setprompt pass:
settitle pw
setok OK
setcancel Cancel
getpin
EOD
)"
		echo "$answers" | grep "^ERR" && return 1
		USER_PW="$( echo "$answers" | grep "^D " | awk '{print $2}' )"
	fi
}
dblask() {
	if [ -t 0 ]; then
		local pass repeated
		read -s -p "$1: " pass || return 1
		printf "\n"
		read -s -p "Repeat: " repeated || return 1
		printf "\n"

		if [[ $pass == "$repeated" ]]; then
			USER_PW="$pass"
		else
			return 1
		fi
	else
		local answers
		answers="$( pinentry-gnome3 <<EOD
setdesc $1
setprompt pass:
settitle pw
setok OK
setcancel Cancel
setrepeat
getpin
EOD
)"
		grep "^ERR" <<< "$answers" && return 1
		USER_PW="$( echo "$answers" | grep "^D " | awk '{print $2}' )"
	fi	
}

decrypt() {
	[[ -d "${TMPDIR}/${FLNAME}" ]] && return 0 #already decrypted from previous
	[[ -f "${ENCDIR}/${FLNAME}" ]] || die "Error: no encrypted library found. Create entries first"
	file -b "${ENCDIR}/${FLNAME}" | grep "GPG.*encrypted" >/dev/null 2>&1 || die "Error: library was not encrypted"

	askkey "Enter decryption key" || die "Decryption key required"
	gpg -d  --pinentry-mode loopback --passphrase "$USER_PW" "${ENCDIR}/${FLNAME}" 2>/dev/null | tar -C "${TMPDIR}" -xj 2>/dev/null || die "bad password"
}

encrypt() {
	[[ -d "${TMPDIR}/${FLNAME}" ]] || die "Error: no information to encrypt"
	[[ -s "${TMPDIR}/${FLNAME}/unlock" ]] && echo "Library unlocked, not saving" && exit 0

	while [[ -z "$USER_PW" ]]; do
		dblask "Enter new master encryption key" || say "key required"
	done
	tar -C "${TMPDIR}" -cj "${FLNAME}" | gpg -c "${GPG_OPTS[@]}" -o "${ENCDIR}/${FLNAME}" --pinentry-mode loopback --passphrase "$USER_PW"
	cleanup
}

lock_open() {
	decrypt
	trap - INT TERM EXIT
	echo "$(date -u)" > "${TMPDIR}/${FLNAME}/unlock" #manual override note
}
remove_lock() {
	rm -f "${TMPDIR}/${FLNAME}/unlock"
}

createlib() {
	dblask "Enter new master encryption key" || die "key required"
	mkdir -p "${TMPDIR}/${FLNAME}"
}

cleanup() {
	[[ -d "${TMPDIR}/${FLNAME}" ]] || exit 0
	[[ -s "${TMPDIR}/${FLNAME}/unlock" ]] && exit 0
	find "${TMPDIR}/${FLNAME}" -type f -exec shred -fzu {} +
	rm -rf "${TMPDIR}/${FLNAME}"
}

check_sneaky_paths() {
	local path
	for path in "$@"; do
		if [[ $path =~ /\.\.$ || $path =~ ^\.\./ || $path =~ /\.\./ || $path =~ ^\.\.$ ]]; then
			die "Error: You've attempted to pass a sneaky path to pass. Go home."
		fi
	done
}

#
# END helper functions
#

#
# BEGIN platform definable
#

# Usage: clip password [username] accountlabel
clip() {
	# This base64 business is because bash cannot store binary data in a shell
	# variable. Specifically, it cannot store nulls nor (non-trivally) store
	# trailing new lines.
	local sleep_argv0="password store sleep on display $DISPLAY"
	pkill -f "^$sleep_argv0" 2>/dev/null && sleep 0.5
	local saved_clip="$(xclip -o -selection clipboard 2>/dev/null | base64)"
	local saved_alt_clip="$(xclip -o -selection primary 2>/dev/null | base64)"
	echo -n "$1" | xclip -selection clipboard || die "Error: Could not copy data to the clipboard"
	if [[ $# -gt 2 ]]; then
		echo -n "$2" | xclip -selection primary || die "Error: Could not copy data to middle mouse clipboard"
	fi
	(
		( exec -a "$sleep_argv0" sleep "$CLIP_TIME" )
		echo "$saved_clip" | base64 -d | xclip -selection clipboard
		[[ $# -gt 2 ]] && echo "$saved_alt_clip" | base64 -d | xclip -selection primary
	) 2>/dev/null & disown
	if [[ $# -eq 2 ]]; then
		say softly "Copied $2 to clipboard. Will clear in $CLIP_TIME seconds."
	else
		say softly "Copied $3 to both clipboards. Will clear in $CLIP_TIME seconds"
	fi
}

#
# END platform definable
#


#
# BEGIN subcommand functions
#

cmd_version() {
	echo "pw v1.7"
}

cmd_usage() {
	cmd_version
	echo
	cat <<-_EOF
	Usage:
	    $PROGRAM [ls] [subfolder]
	        List passwords.
	    $PROGRAM find pass-names...
	    	List passwords that match pass-names.
	    $PROGRAM [show] [--clip[=line-number],-c[line-number]] pass-name
	        Show existing password and optionally put it on the clipboard.
	        If put on the clipboard, it will be cleared in $CLIP_TIME seconds.
	    $PROGRAM grep search-string
	        Search for password files containing search-string when decrypted.
	    $PROGRAM insert [--echo,-e | --multiline,-m] [--force,-f] pass-name
	        Insert new password. Optionally, echo the password back to the console
	        during entry. Or, optionally, the entry may be multiline. Prompt before
	        overwriting existing password unless forced.
	    $PROGRAM edit pass-name
	        Insert a new password or edit an existing password using ${EDITOR:-vi}.
	    $PROGRAM generate [--no-symbols,-n] [--clip,-c] [--in-place,-i | --force,-f] pass-name [pass-length]
	        Generate a new password of pass-length (or $GENERATED_LENGTH if unspecified) with optionally no symbols.
	        Optionally put it on the clipboard and clear board after $CLIP_TIME seconds.
	        Prompt before overwriting existing password unless forced.
	        Optionally replace only the first line of an existing file with a new password.
	    $PROGRAM rm [--recursive,-r] [--force,-f] pass-name
	        Remove existing password or directory, optionally forcefully.
	    $PROGRAM mv [--force,-f] old-path new-path
	        Renames or moves old-path to new-path, optionally forcefully
	    $PROGRAM cp [--force,-f] old-path new-path
	        Copies old-path to new-path, optionally forcefully
	    $PROGRAM open [--yes,-y] [-t,--timeout SECONDS]
	    	Decrypts password library and leaves open. Leaves your data un-protected. Very dangerous. Optionally set a timeout to auto close. Use 0 for forever (extra dangerous)
	    $PROGRAM close [--save,-s]
	    	Cleans up an open library, optionally saving any changes made while open. Otherwise just securely deletes.
	    $PROGRAM help
	        Show this text.
	    $PROGRAM version
	        Show version information.

	More information may be found in the pass(1) man page.
	_EOF
}

cmd_list() {
	local path="$1"
	check_sneaky_paths "$path"
	decrypt
	tree -C -l --noreport "${TMPDIR}/${FLNAME}/${path}" | tail -n +2 | sed -E 's/\.txt(\x1B\[[0-9]+m)?( ->|$)/\1\2/g' #remove .txt at end of line, but keep colors
}

cmd_show() {
	local opts clip_location clip=0
	opts="$(getopt -o c:: -l clip:: -n "$PROGRAM" -- "$@")"
	local err=$?
	eval set -- "$opts"
	while true; do case $1 in
		-c|--clip) clip=1; clip_location="${2:-1}"; shift 2 ;;
		--) shift; break ;;
	esac done
	clip_location="${clip_location:-1}"

	[[ $err -ne 0 ]] && die "Usage: $PROGRAM $COMMAND [--clip[=line-number],-c[line-number]] [pass-name]"

	local path="$1"
	local passfile="${TMPDIR}/${FLNAME}/$path.txt"
	check_sneaky_paths "$path"

	decrypt
	if [[ -d ${passfile%.txt} ]]; then # "show" a directory by listing contents
		COMMAND="list"
		cmd_list "$path"
	elif [[ -f $passfile ]]; then
		if [[ $clip -eq 0 && -t 0 ]]; then
			cat "$passfile"
		else
			[[ $clip_location =~ ^[0-9]+$ ]] || die "Clip location '$clip_location' is not a number."
			local pass="$(tail -n +${clip_location} "$passfile" | head -n 1)"
			local pass_alt="$(tail -n +${clip_location} "$passfile" | head -n 2 | tail -n 1 )"
			[[ -n $pass ]] || die "There is no password to put on the clipboard at line ${clip_location}."
			if [[ -n "$pass_alt" && "$pass_alt" != "\n" ]]; then
				clip "$pass" "$pass_alt" "$path"
			else
				clip "$pass" "$path"
			fi
		fi
	elif [[ -d $PREFIX/$path ]]; then
		cmd_list "$path"
	else
		die "Error: $path was not found."
	fi
}

cmd_find() {
	[[ -z "$@" ]] && die "Usage: $PROGRAM $COMMAND pass-names..."
	decrypt
	IFS="," eval 'echo "Search Terms: $*"'
	local terms="*$(printf '%s*|*' "$@")"
	tree -C -l --noreport -P "${terms%|*}" --prune --matchdirs --ignore-case "${TMPDIR}/${FLNAME}" | tail -n +2 | sed -E 's/\.txt(\x1B\[[0-9]+m)?( ->|$)/\1\2/g'
}

cmd_grep() {
	[[ $# -ne 1 ]] && die "Usage: $PROGRAM $COMMAND search-string"
	decrypt
	grep --color=always -H -nr "$1" "${TMPDIR}/${FLNAME}" | sed "s@${TMPDIR}/${FLNAME}/@@g" # strip out leading 
}

cmd_insert() {
	if [ ! -f "${ENCDIR}/${FLNAME}" ]; then
		if approve "password library not found. Create a new one?"; then
			createlib
		else
			exit 1
		fi
	else
		decrypt
	fi

	local opts multiline=0 force=0
	opts="$(getopt -o mf -l multiline,force -n "$PROGRAM" -- "$@")"
	local err=$?
	eval set -- "$opts"
	while true; do case $1 in
		-m|--multiline) multiline=1; shift ;;
		-f|--force) force=1; shift ;;
		--) shift; break ;;
	esac done

	[[ $err -ne 0 || $# -ne 1 ]] && die "Usage: $PROGRAM $COMMAND [--multiline,-m] [--force,-f] pass-name"
	local path="${1%/}"
	local passfile="${TMPDIR}/${FLNAME}/${path}.txt"
	check_sneaky_paths "$path"

	if [[ $force -eq 0 && -e $passfile ]]; then
		approve "An entry already exists for $path. Overwrite it?" || exit 1
	fi

	mkdir -p -v "${TMPDIR}/${FLNAME}/$(dirname "$path")"

	if [[ $multiline -eq 1 ]]; then
		$EDITOR "$passfile"
		if [[ -s $passfile ]]; then
			encrypt
		elif [[ -f $passfile ]]; then
			die "Not saved. Empty password"
		else
			die "Not saved. Aborted by user"
		fi
	else
		local passcache gotpass
		[[ -n "$USER_PW" ]] && passcache="$USER_PW"
		if askkey "Enter password for $path"; then
			echo "$USER_PW" > "$passfile"
			[[ -n "$passcache" ]] && USER_PW="$passcache"
			encrypt
		else
			[[ -n "$passcache" ]] && USER_PW="$passcache"
			die "Empty password. Aborted."
		fi
	fi
}

cmd_edit() {
	[[ $# -ne 1 ]] && die "Usage: $PROGRAM $COMMAND pass-name"

	local path="${1%/}"
	check_sneaky_paths "$path"

	decrypt
	local passfile="${TMPDIR}/${FLNAME}/${path}.txt"
	[[ -f $passfile ]] || die "${path} not found"
	local before="$(md5sum $passfile | awk '{print $1}')"
	$EDITOR $passfile

	if [[ -s $passfile ]]; then
		local after="$(md5sum $passfile | awk '{print $1}')"
		if [[ $before == "$after" ]]; then
			say "Entry not changed"
		else
			encrypt
		fi
	elif [[ -f $passfile ]]; then
		die "${path} cannot have empty password. Remove instead. Aborting edit"
	else
		die "Cannot remove ${path} this way. Use rm command. Aborting"
	fi
}

cmd_generate() {
	local opts clip=0 force=0 symbols="-y" inplace=0
	opts="$(getopt -o ncif -l no-symbols,clip,in-place,force -n "$PROGRAM" -- "$@")"
	local err=$?
	eval set -- "$opts"
	while true; do case $1 in
		-n|--no-symbols) symbols=""; shift ;;
		-c|--clip) clip=1; shift ;;
		-f|--force) force=1; shift ;;
		-i|--in-place) inplace=1; shift ;;
		--) shift; break ;;
	esac done

	[[ $err -ne 0 || ( $# -ne 2 && $# -ne 1 ) || ( $force -eq 1 && $inplace -eq 1 ) ]] && die "Usage: $PROGRAM $COMMAND [--no-symbols,-n] [--clip,-c] [--in-place,-i | --force,-f] pass-name [pass-length]"
	local path="$1"
	local length="${2:-$GENERATED_LENGTH}"
	check_sneaky_paths "$path"
	[[ ! $length =~ ^[0-9]+$ ]] && die "Error: pass-length \"$length\" must be a number."
	mkdir -p -v "$PREFIX/$(dirname "$path")"
	local passfile="$PREFIX/$path.gpg"

	[[ $inplace -eq 0 && $force -eq 0 && -e $passfile ]] && approve "An entry already exists for $path. Overwrite it?" || exit 1

	local pass="$(pwgen -s $symbols $length 1)"
	[[ -n $pass ]] || exit 1
	if [[ $inplace -eq 0 ]]; then
		$GPG -e -o "$passfile" "${GPG_OPTS[@]}" <<<"$pass" || die "Password encryption aborted."
	else
		local passfile_temp="${passfile}.tmp.${RANDOM}.${RANDOM}.${RANDOM}.${RANDOM}.--"
		if $GPG -d "${GPG_OPTS[@]}" "$passfile" | sed $'1c \\\n'"$(sed 's/[\/&]/\\&/g' <<<"$pass")"$'\n' | $GPG -e -o "$passfile_temp" "${GPG_OPTS[@]}"; then
			mv "$passfile_temp" "$passfile"
		else
			rm -f "$passfile_temp"
			die "Could not reencrypt new password."
		fi
	fi
	local verb="Add"
	[[ $inplace -eq 1 ]] && verb="Replace"
	git_add_file "$passfile" "$verb generated password for ${path}."

	if [[ $clip -eq 0 ]]; then
		printf "\e[1m\e[37mThe generated password for \e[4m%s\e[24m is:\e[0m\n\e[1m\e[93m%s\e[0m\n" "$path" "$pass"
	else
		clip "$pass" "$path"
	fi
}

cmd_delete() {
	local opts recursive="" force=0
	opts="$(getopt -o rf -l recursive,force -n "$PROGRAM" -- "$@")"
	local err=$?
	eval set -- "$opts"
	while true; do case $1 in
		-r|--recursive) recursive="-r"; shift ;;
		-f|--force) force=1; shift ;;
		--) shift; break ;;
	esac done
	[[ $# -ne 1 ]] && die "Usage: $PROGRAM $COMMAND [--recursive,-r] [--force,-f] pass-name"
	local path="$1"
	check_sneaky_paths "$path"

	local passdir="${TMPDIR}/${FLNAME}/${path%/}"
	local passfile="${TMPDIR}/${FLNAME}/${path}.txt"

	decrypt

	[[ -f $passfile && -d $passdir && $path == */ || ! -f $passfile ]] && passfile="$passdir"
	[[ -e $passfile ]] || die "Error: $path is not in the password store."

	[[ $force -eq 1 ]] || approve "Are you sure you would like to delete $path?" || exit 1

	rm $recursive -f -v "$passfile"
	rmdir -p "${passfile%/*}" 2>/dev/null || true

	encrypt
}

cmd_copy_move() {
	local opts move=1 force=0
	[[ $1 == "copy" ]] && move=0
	shift
	opts="$(getopt -o f -l force -n "$PROGRAM" -- "$@")"
	local err=$?
	eval set -- "$opts"
	while true; do case $1 in
		-f|--force) force=1; shift ;;
		--) shift; break ;;
	esac done
	[[ $# -ne 2 ]] && die "Usage: $PROGRAM $COMMAND [--force,-f] old-path new-path"
	check_sneaky_paths "$@"

	decrypt

	local old_path="${TMPDIR}/${FLNAME}/${1%/}"
	local old_dir="$old_path"
	local new_path="${TMPDIR}/${FLNAME}/$2"

	if ! [[ -f ${old_path}.txt && -d $old_path && $1 == */ || ! -f ${old_path}.txt ]]; then
		old_dir="${old_path%/*}"
		old_path="${old_path}.txt"
	fi
	echo "$old_path"
	[[ -e $old_path ]] || die "Error: $1 is not in your library."

	mkdir -p -v "${new_path%/*}"
	[[ -d $old_path || -d $new_path || $new_path == */ ]] || new_path="${new_path}.txt"

	local interactive="-i"
	[[ ! -t 0 || $force -eq 1 ]] && interactive="-f"

	if [[ $move -eq 1 ]]; then
		mv $interactive -v "$old_path" "$new_path" || exit 1
		rmdir -p "$old_dir" 2>/dev/null || true
		encrypt
	else
		cp $interactive -r -v "$old_path" "$new_path" || exit 1
		encrypt
	fi
}

cmd_update() {
	decrypt
	unset USER_PW
	encrypt
}


cmd_open() {
	local skip=0 pw_timeout=10 opts="$(getopt -o yt: -l yes,timeout: -n "$PROGRAM" -- "$@")"
	eval set -- "$opts"
	while true; do case $1 in
		-y|--yes) skip=1; shift ;;
		-t|--timeout) pw_timeout="$2"; shift 2;;
		--) shift; break ;;
	esac done
	[[ "$skip" -eq 1 ]] || approve "Leaving your password library decrypted in incredibly unsafe. You are responsible for removing it. Continue?" || exit 1
	lock_open

	if [[ $pw_timeout -eq 0 ]]; then
		say softly "Library unlocked. It will not auto-close. Please '$PROGRAM close' when you are finished"
	else
		local sleep_argv0="pw-autoclose"
		#pkill -f "^$sleep_argv0" 2>/dev/null
		(
			( exec -a "$sleep_argv0" sleep "$pw_timeout" )
			remove_lock
			cleanup
		) 2>/dev/null & disown

		say softly "Library unlocked. Will auto-close in $pw_timeout seconds"
	fi


}

cmd_close() {
	local save=0 opts="$(getopt -o s -l save -n "$PROGRAM" -- "$@")"
	eval set -- "$opts"
	while true; do case $1 in
		-s|--save) save=1; shift ;;
		--) shift; break ;;
	esac done

	if [[ $save -eq 1 ]]; then
		remove_lock
		encrypt
	else
		remove_lock
		cleanup
	fi
}

#
# END subcommand functions
#

trap cleanup INT TERM EXIT

PROGRAM="${0##*/}"
COMMAND="$1"

case "$1" in
	help|-h|--help) shift;			cmd_usage "$@" ;;
	version|-v|--version) shift;	cmd_version "$@" ;;
	ls|list) shift;					cmd_list "$@" ;;
	show) shift;					cmd_show "$@" ;;
	find|search) shift;				cmd_find "$@" ;;
	grep) shift;					cmd_grep "$@" ;;
	insert|add) shift;				cmd_insert "$@" ;;
	edit) shift;					cmd_edit "$@" ;;
	generate) shift;				cmd_generate "$@" ;;
	delete|rm|remove) shift;		cmd_delete "$@" ;;
	rename|mv) shift;				cmd_copy_move "move" "$@" ;;
	copy|cp) shift;					cmd_copy_move "copy" "$@" ;;
	update|up) shift;				cmd_update "$@" ;;
	open|op|unlock|decrypt) shift;	cmd_open "$@" ;;
	close|cl|lock|encrypt) shift;	cmd_close "$@" ;;
	*) COMMAND="show";				cmd_show "$@" ;;
esac
exit 0
